### ISSUE TITLE
support for setup.py register command is broken

### ISSUE THREAD

**Issue by petri** [pypiserver/pypiserver#91]
I have fresh pypiserver 1.1.7 running on current Debian Stable, on Python2.7.

For some reason, running "python setup.py sdist register upload" works, but just "setup.py register" does not (locally):

    Server response (500): <urlopen error [Errno 104] Connection reset by peer>

Or from a remote OSX system:

    Server response (500): <urlopen error [Errno 32] Broken pipe>

... and likewise, using jarn.mkrelease (tool to automate releases) from the OSX system to do uploads fails:
(it does register - sdist - upload in three separate HTTP calls, I guess)

    running register
    Registering z3c.form to http://pypi.vidamin.com:8765
    Server response (500): <urlopen error [Errno 32] Broken pipe>
    ERROR: register failed

In all these cases, the pypiserver log shows just:

    "POST / HTTP/1.1" 401 713 "-" "Python-urllib/2.7"

Which is not very helpful...

pypiserver should accept the register call on its own, even if just to satisfy clients (and users) that try to do "setup.py register" on its own without other commands. My suggestion would be to simply silently accept the call, and do nothing (except log the event perhaps).

Note: I suspect this issue may be behind a few other issue reports about "random" 401 or 500 errors, as well... I thought for a long time I was seeing the same issue.

At the very least, it should be documented that doing just "setup.py register" does not work.

**Comment by ankostis** [comment]
@petri: cannot reproduce :-(

I tried it on my linux-py3.4 with the latest 1.8.1-beta.1, both with a password-file and without (the newly ``-P .`` option) and both commands with `register worked.

Here it is my `~/.pypirc`

	[distutils]
	index-servers =
	  #pypi
	  internal
	
	[internal]
	repository: http://localhost:8080
	username: foo
	password: bar

Server start-up commands:

     pypi-server -p 8080 -P . -a ''  ./p & ## Un-authed uploads (password ignored)
     pypi-server -p 8080 -P htpassd ./p &

Register & Upload commands:

     python setup.py  register -r internal
     python setup.py  register -r internal  sdist upload -r internal

**Comment by ankostis** [comment]
@petri would you mind confirming/closing this?

**Comment by petri** [comment]
I don't know what to say. For me, ```python setup.py register still ``` fails about eight times out of nine. I have not tried the latest beta though.

Here's how I start pypiserver:

```/opt/pypiserver/bin/python ./bin/pypi-server -a download,update -P .htpasswd -p 8765 packages```

I am running pypiserver on a fast VPS that's not far, and there's good connectivity in-between.

**Comment by ankostis** [comment]
But when you invoke `register` without `-r` option it most probably registers to the central PyPi (unless otherwise instructed by your `~/.pypirc`), correct?

I cannot judge for sure, because I'm still missing your exact register-command and your `~/.pypirc` file,  but the correct command would be:

    python setup.py bdist_wheel register -r local upload -r local 

In any case, to avoid all this complications, do not use the`register` command at all! 
Just follow the instructions in the `README` and modify your `~/.pypirc` with an editor (that's what the register-cmd does) . 
Even better, use the [twine](https://pypi.python.org/pypi/twine) library to avoid storing your passwords to disk.

**Comment by petri** [comment]
Thanks for the effort - sorry, I was imprecise earlier, I was of course using the -r option... I've also tried pretty much the every variation of register with sdist/bdist/wheel/whatever + upload to see if that makes a difference. Yes, working without register command is a partial workaround, but the issue remains when using release tools that have hardcoded behavior of running register + something else.

For what it's worth, here's my pypirc, with just passwords etc. obfuscated:

```
[distutils]
index-servers =
  mycompany
  pypi

[pypi]
username:me
password:mypass

[mycompany]
repository: http://my.company.com:1234
username: user
password: password
```

**Comment by Tythos** [comment]
Very interesting bug. It was causing me issues, too, and I didn't want to neglect the register operation as I am hoping to capture it in an extended pypiserver application. Turns out to be a single-character fix--but it took me hours to trace through the request/response headers, and into the depths of distutils and urllib2, with a brief misdirected detour through Apache wsgi proxy-forwarding...

**Comment by petri** [comment]
Hm I applied the fix to my install of pypiserver and now when doing "python setup.py register -r mypypi" I see mostly "error: [Errno 54] Connection reset by peer" and sometimes also "Server response (401): Unauthorized".

**Comment by Tythos** [comment]
I can't guarantee a fix that wasn't with my version--1.1.9--but I can tell you some of the steps I took to debug the issue:

* Drop some debug prints in distutils.request and urllib2 to look at the exact headers that are being returned. Are you getting a www-authenticate response? Compare headers with those from 'upload'.
* Ask OpenSSL to look at the response you are getting from the server:

> openssl s_client -connect my.server.com:443

* The "connection reset" issue makes me wonder how many threads you have allocated. How are you starting and hosting your WSGI application?
* Look at the specific request and response headers on the server side before they are processed and returned, respectively. The specific steps will depend on your server configuration (i.e., raw WSGI vs .Apache vs. nginx, etc.)

**Comment by petri** [comment]
Well, to make things as simple as possible, I ran pypiserver as:

    ./bin/pypi-server -a download,update -P .htpasswd -p 8765 packages

And applied the fix you suggest. I then tried logging in using just a browser, and could not. Removing the fix allowed me to log in again.

**Comment by Tythos** [comment]
That's very strange. Can you look at your bottle.py and verify in HTTPResponse the right name of the argument is 'headers'?

**Comment by petri** [comment]
Taking a look at bottle.py inside latest published pypiserver 1.1.8 vs what's inside master, ie. 1.1.9, I see the HTTPResponse init has changed.

From Bottle 0.11.6:
```
class HTTPResponse(Response, BottleException):
    def __init__(self, body='', status=None, header=None, **headers):
        if header or 'output' in headers:
            depr('Call signature changed (for the better)')
            if header: headers.update(header)
            if 'output' in headers: body = headers.pop('output')
        super(HTTPResponse, self).__init__(body, status, **headers)
```

To Bottle 0.13-dev:

```
class HTTPResponse(Response, BottleException):
    def __init__(self, body='', status=None, headers=None, **more_headers):
        super(HTTPResponse, self).__init__(body, status, headers, **more_headers)

So from what I can tell, the fix is needed for 1.1.9 to work correctly with the upgraded bottle.py.

**Comment by petri** [comment]
Ok @Tythos I cloned master from your fork and I am happy to say I have not seen any of the issues I've described here :)

**Comment by petri** [comment]
This can be closed as soon as #98 is merged.

**Comment by ankostis** [comment]
I want to deeply thank you for your efforts, and apologize for not having the time to release.

Given this opportunity, please feel free anyone of you to ask for access into the dev-team of this project.
